xdsl/pattern_rewriter.py

```diff
@@ -609,6 +609,52 @@ def match_and_rewrite(self, op: Operation, rewriter: PatternRewriter) -> None:
         return
 
 
+@dataclass(eq=False)
+class Worklist:
+    _op_stack: list[Operation | None] = field(default_factory=list, init=False)
+    """
+    The list of operations to iterate over, used as a last-in-first-out stack.
+    Operations are added and removed at the end of the list.
+    Operation that are `None` are meant to be discarded, and are used to
+    keep removal of operations O(1).
+    """
+
+    _map: dict[Operation, int] = field(default_factory=dict, init=False)
+    """
+    The map of operations to their index in the stack.
+    It is used to check if an operation is already in the stack, and to
+    remove it in O(1).
+    """
+
+    def is_empty(self) -> bool:
+        """Check if the worklist is empty."""
+        return bool(self._op_stack)
+
+    def push(self, op: Operation):
+        """
+        Push an operation to the end of the worklist, if it is not already in it.
+        """
+        if op not in self._map:
+            self._map[op] = len(self._op_stack)
+            self._op_stack.append(op)
+
+    def pop(self) -> Operation | None:
+        """Pop the operation at the end of the worklist."""
+        while self._op_stack:
+            op = self._op_stack.pop()
+            if op is not None:
+                del self._map[op]
+                return op
+        return None
```

Comment:
It feels weird for the stack to be empty twice, effectively, first explicitly, and then by checking that the op is not `None`. Maybe something like this?
```suggestion
        if not self._op_stack:
            return None
        op = self._op_stack[-1]
        del self._op_stack[-1]
        del self._map[op]
        return op
```

Reply:
I'm not sure I understand your comment here.
The reason why there is this while loop is because any `None` value in our `_op_stack` is a value that was deleted.
Instead of removing it from the stack, we rename it to `None`, so we can essentially remove them in `O(1)` instead of `O(n)`.
So this removes all values that are `None` at the end.

Reply:
Ah, could you please make that a little clearer in the docstrings? I completely missed the whole `None` mechanism. It feels like a good idea to first find the index of the last non-`None` value in the list, remove all the values after that index, and then pop, instead of looping.

Reply:
It is in the docstring, but in the `_op_stack` one ;)
I'll add a comment on top of this part to make it clearer, and indeed should first look at it then delete it!

Reply:
Actually, I'm not sure that I should first look for the last non-`None`, then remove all at once.

Since most of the time, there are no `None`. This only happens when you remove an operation that is not the matched one, so that's pretty rare.

First looking at the last non-`None` value will be slower, as we currently do only one `pop` and one `if isinstance(.., None)` in the "normal" case, while checking for `None` might do multiple checks. 

Also, I would argue that the current version is clearer and less error-prone?

Reply:
I see what you mean. Sounds good :)

Reply:
Would still be good to have more doc strings and comments

Reply:
Just to be sure, what other doc strings are you referring to?
I'm adding comments there, but I'm not sure what doc strings should be added?

Reply:
The comment you added explains it well, thank you for adding it